### ISSUE TITLE
Closing dialog not visible/reachable

### ISSUE THREAD

**Issue by farosch** [notepad-plus-plus/notepad-plus-plus#7646]
### Description of the Issue

The following closing dialog seems to always appear centered to its' parent.
![image](https://user-images.githubusercontent.com/22905106/69478207-3cb49080-0df0-11ea-8121-6488aeac5fe7.png)
This may lead to the dialog not being visible/usable, when the main window is mainly outside of the visible area.

### Steps to Reproduce the Issue

1. Open a new file
2. Make changes to the file
3. Move np++ window in a way that most of the application is outside of the desktop
4. Close tab containing file
5. Closing dialog takes focus and is not visible

### Expected Behavior
Closing dialog should be within visible bounds.
### Actual Behavior
Closing dialog is **not** within visible bounds.

### Debug Information

*none needed*

**Comment by sasumner** [comment]
>none needed

Did you read the template?:

```
### Debug Information
<!--- Debug Info can be found under under the "?" menu -->
```

So you go to the `?` menu in Notepad++ and there's an entry called `Debug Info...` and you invoke that and then you copy the results to the clipboard and you paste the results here.

BUT...

One presumes you are talking about Notepad++ 7.8.1, with the newer style Save dialog (the one with the additional  `Yes to All` and `No to All` buttons).

Indeed, I can confirm this behavior.  If I move the non-maximized Notepad++ main window to my rightmost monitor such that, say, only the `F` in the `File` menu is visible, and then attempt to close a modified file using `File` > `Close`, then I see no Save dialog as it is created centered to the window, which places it entirely offscreen.  As it is modal, attempting to interact with the Notepad++ main window with the mouse is unsuccessful; one recovery mechanism is to press `Esc` once to cancel the Save dialog.

Attempting to replicate this on an earlier (I used 7.6.6) version (with the 3-button Yes/No/Cancel) Save dialog, it doesn't replicate; the Save dialog is created centered on my right-most monitor.

Suggested resolution:  Center on N++ main window only if the Save dialog will be onscreen in its entirety; otherwise, center on the monitor that the largest portion of the N++ main window appears on.

**Comment by xylographe** [comment]
Centring is implemented through [StaticDialog::goToCenter()](https://github.com/notepad-plus-plus/notepad-plus-plus/blob/f44c3579994e0259c2f5ce24cf41555b92a0d005/PowerEditor/src/WinControls/StaticDialog/StaticDialog.cpp#L65).

**Comment by sasumner** [comment]
@xylographe How does that centering help?  The code centers it on the parent, which is exactly the _problem_ in this case (even without that call being made).

**Comment by sasumner** [comment]
So we are probably in a no-win situation here.  Removing the call to `goToCenter` from the `DoSaveOrNotBox::run_dlgProc`'s `WM_INITDIALOG` section results in a solution to this issue (save dialog will appear fully on-screen even when most of the main N++ window is off-screen), BUT for the normal case (N++ main window fully or mostly visible) users will notice and probably complain that the save dialog is not centered on the main window.

**Comment by sasumner** [comment]
**Replacing the aforementioned `goToCenter()` call** with the following code block might be a good compromise. 

Without actually centering the dialog first, we check to see if the save dialog will be partially (or fully) off-screen if we were to actually center it; if that is true then we don't center it ourselves but rather let Windows handle the placement.  

The reason we don't let Windows handle it all the time is that it picks a somewhat arbitrary placement (I'm sure it isn't truly arbitrary -- and makes sense to Windows itself!) when a centered save dialog looks a whole lot better.

```
RECT parent_rc;
::GetClientRect(_hParent, &parent_rc);
POINT parent_center;
parent_center.x = parent_rc.left + (parent_rc.right - parent_rc.left) / 2;
parent_center.y = parent_rc.top + (parent_rc.bottom - parent_rc.top) / 2;
::ClientToScreen(_hParent, &parent_center);
POINT centered_dlg_top_left, centered_dlg_bottom_right;
centered_dlg_top_left.x = parent_center.x - (_rc.right - _rc.left) / 2;;
centered_dlg_top_left.y = parent_center.y - (_rc.bottom - _rc.top) / 2;;
centered_dlg_bottom_right.x = centered_dlg_top_left.x + (_rc.right - _rc.left);
centered_dlg_bottom_right.y = centered_dlg_top_left.y + (_rc.bottom - _rc.top);
if ((MonitorFromPoint(centered_dlg_top_left, MONITOR_DEFAULTTONULL) == NULL) || 
    (MonitorFromPoint(centered_dlg_bottom_right, MONITOR_DEFAULTTONULL) == NULL))
{
    // If we center dialog it will be partially or fully off-screen;
    // so DON'T center it, let Windows place it because it will keep it fully on-screen
}
else
{
    goToCenter();  // center dialog on parent window
}
```

**Comment by sasumner** [comment]
Probably better for those "odd" monitor arrangements to check _all four_ corners of the potentially-centered dialog:

```
RECT parent_rc;
::GetClientRect(_hParent, &parent_rc);
POINT parent_center;
parent_center.x = parent_rc.left + (parent_rc.right - parent_rc.left) / 2;
parent_center.y = parent_rc.top + (parent_rc.bottom - parent_rc.top) / 2;
::ClientToScreen(_hParent, &parent_center);
POINT corners[4];
/*
[0]+-----------------------+[1]
...|                       |
...|                       |
[2]+-----------------------+[3]
*/
corners[0].x = parent_center.x - (_rc.right - _rc.left) / 2;;
corners[0].y = parent_center.y - (_rc.bottom - _rc.top) / 2;;
corners[3].x = corners[0].x + (_rc.right - _rc.left);
corners[3].y = corners[0].y + (_rc.bottom - _rc.top);
corners[1].x = corners[3].x; 
corners[1].y = corners[0].y;
corners[2].x = corners[0].x;
corners[2].y = corners[3].y;
int i;
for (i = 0; i < 4; ++i)
{
    if (::MonitorFromPoint(corners[i], MONITOR_DEFAULTTONULL) == NULL)
    {
        // a corner of a parent-centered dialog would not be visible on any monitor
        break;
    }
}
if (i == 4)
{
    // all 4 corners of (a centered) dialog would be visible on screen, so go ahead and center it:
    goToCenter();
    // (otherwise, just let Windows handle the placement, because its choice will be "reasonable")
}

```

**Comment by xylographe** [comment]
@sasumner 
This also affects other dialogues. For example, if I move the NPP window all the way down until only the top is visible, and then press **F1**, the About dialogue opens, but is not visible ( pressing enter closes the invisible dialogue). Hence, I would suggest integrating the guarding code block into **`goToCenter`** itself, where all dialogues can take advantage.

**Comment by sasumner** [comment]
Yes, I searched for all existing occurrences of `goToCenter()` and tried them all out.  I saw what you saw for the "About" box, but some of the others had interesting (i.e., different) outcomes.  Some cause the dialog to appear centered on the desktop, which probably means (didn't investigate) that the N++ main window is not their parent (maybe they have no parent).

Anyway, I agree this should all be unified, and thank you for your response; I was hoping someone would offer an opinion.  That's why I put some code here rather than in a PR.

However, we often see a "bigger" code change rejected, so at this point I'm not sure how exactly I want to proceed.

To me, and probably others, this is a very minor issue.  I just jumped on it because I thought it might be interesting, not particularly because it needed solving. :-)

**Comment by sasumner** [comment]
A method for making sure the "Find" window doesn't get "lost" out of user view was presented in the implementation of #905 .  Perhaps something similar could be done for this issue.

**Comment by sasumner** [comment]
Related:  https://community.notepad-plus-plus.org/topic/18861/save-prompt-off-screen
